pkg/util/fastjson/extract.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fastjson
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+)
+
+type topLevelJSONTokenIter struct {
+	d     *json.Decoder
+	level int
+}
+
+func newTopLevelJSONTokenIter(content []byte) *topLevelJSONTokenIter {
+	d := json.NewDecoder(bytes.NewReader(content))
+	d.UseNumber()
+	return &topLevelJSONTokenIter{
+		d: d,
+	}
+}
+
+func unexpectedEOF(err error) error {
+	if err == io.EOF {
+		return io.ErrUnexpectedEOF
+	}
+	return err
+}
+
+// next iterates and returns the top-level key and value as token. The odd
+// invocations return key tokens, the even invocations return value tokens. If
+// the value is a JSON array or object, all tokens belongs to the value will be
+// returned.
+func (i *topLevelJSONTokenIter) next(discard bool) ([]json.Token, error) {
```

Comment:
maybe split it into 2 methods to better fit what this struct named as
- nextKey
- readOrDiscardValue

Reply:
https://github.com/pingcap/tidb/pull/55304/commits/80ec0cb88843796d33da48cc20f3ddb98a383278